### ISSUE TITLE
VW JNI should convert signals to exceptions so the JVM can recover.

### ISSUE THREAD

**Issue by deaktator** [JohnLangford/vowpal_wabbit#734]
# Problem

Currently, as of `"com.github.johnlangford" % "vw-jni" % "8.0.0"`, signals produced in the native VW library are unrecoverable from within the JVM code which calls the JNI VW library.  **These signals should be turned into exceptions** so that the JVM doesn't *have to* crash.

## Possible solutions

- Use a library like [CoffeeCatch](https://github.com/xroche/coffeecatch) for conversion
- Manually convert using [signal](https://gist.github.com/jvranish/4441299) and [sigaction](http://www.cplusplus.com/forum/unices/16430/) in [&lt;signal.h&gt;](http://pubs.opengroup.org/onlinepubs/009695399/basedefs/signal.h.html)

### Unit test not only fails, but crashes the JVM
```scala
// Scala Code
import org.apache.commons.vfs2
import org.junit.Test
class CorruptedModelTest {
  @Test def test(): Unit = {
    // Points to a corrupted model.
    val badModel = vfs2.VFS.getManager.resolveFile("res:bad_model.vw").getName.getPath
    println(s"attempting to instantiate corrupted VW model: $badModel")
    // This try won't do anything because this is currently a SIGSEGV 
    // so the JVM will just blow up.
    try { 
      // This will fail and the JVM cannot recover.
      new VW(s"--quiet -i $badModel") 
    }
    catch {
      // If we could successful trap this error, like we should be able to, 
      // the test would pass.
      case trapThisError: Throwable =>
    }
  }
}
```

### Unit test error message

```
Running CorruptedModelTest
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fff957243a0, pid=18585, tid=4867
#
# JRE version: Java(TM) SE Runtime Environment (7.0_60-b19) (build 1.7.0_60-b19)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (24.60-b09 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libsystem_platform.dylib+0x13a0]  _platform_memmove$VARIANT$Nehalem+0x60
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/xyz/Documents/code/project/hs_err_pid18585.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.sun.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
/bin/sh: line 1: 18585 Abort trap: 6           /Library/Java/JavaVirtualMachines/jdk1.7.0_60.jdk/Contents/Home/jre/bin/java -jar /Users/xyz/Documents/code/project/target/surefire/surefirebooter4818290815839838654.jar /Users/xyz/Documents/code/project/target/surefire/surefire5145374967874691863tmp /Users/xyz/Documents/code/project/target/surefire/surefire_01068200045152922901tmp
```

**Comment by JohnLangford** [comment]
This seems like a good idea.  This was implemented in the new C# interface
by having a C++-side function catch the signal and rethrow it properly for
the C# side which I believe is your option 2.

-John

**Comment by jmorra** [comment]
I agree, however, I'm not sure how to implement this as this is already happening on the (C++)[https://github.com/JohnLangford/vowpal_wabbit/blob/master/java/src/main/c%2B%2B/vw_VW.cc] side but this particular exception isn't being caught from the Java_vw_VW_initialize function.  If you can point me to somewhere that shows how this is done then I'll absolutely update this.

**Comment by JohnLangford** [comment]
Hmm, maybe I'm mistaken.  I think Ryan is referring to unix-style signals.
Ryan, what is a sequence of commands that exhibits the issue?

-John

**Comment by deaktator** [comment]
The way that I did it above was to use `--quiet -i badModel.vw` where `badModel.vwl` points to a model I created and then intentionally corrupted.  I used something like the following to create the model: 

```bash
(paste -d '\n' <(jot -b "-1 |" 100) \
               <(jot -b "1 |" 100)) \
| vw --quiet \
     --link logistic \
     --loss_function logistic \
     -f badModel.vw
```

Then I manually corrupted the file.

**Comment by deaktator** [comment]
BTW, I *was* talking about unix-style (POSIX) signals.

**Comment by jmorra** [comment]
I've done a little more research on this and here's what I've uncovered.  When I supply an old VW model (in my case one trained on some version 7.8) the above test will pass.  This is good as it'll catch an error where the user attempts to load an old model and will throw the error in the Java error.  The condition still exists though that when a corrupted file is attempted to be loaded the program segfaults and this is not catchable by the try/catch syntax on the C side.  Here is some output to illustrate what I'm seeing.

```
Mac:~/git/vowpal_wabbit/java$vw -i src/test/resources/vw_bad.model 
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
Segmentation fault: 11
Mac:~/git/vowpal_wabbit/java$vw -i src/test/resources/vw_7.8.model 
vw (./io_buf.h:204): bad model format!
```

**Comment by JohnLangford** [comment]
Right.  Essentially, we need to either (a) install a signal handler to make
this behave properly or (b) fix the loading code so that it cannot
segfault.  An example of this is provided here:

https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/parser.cc#L133
and here:
https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/parser.cc#L512
(for daemon mode)

We could modify things so the signal handler raises an exception when not
in daemon mode.

-John

**Comment by deaktator** [comment]
Exactly right @JohnLangford. Either is sufficient.  We can fix the entire thing or make the loading code recoverable from certain signals.  

I think that if the former is chosen, it would be a huge departure from the current way VW does it.  Therefore, it is a rather big, breaking change.  Imagine anyone that calls the VW executable via a wrapper shell script and does the error handling in the script.  The third-party error handling logic would have to change to accommodate exceptions instead of signals.  While I personally prefer this approach, I think it is "*high risk*".

To mitigate the risk, there could be to have static / global code that captures signals and emits exceptions that can be both directly callable from a C library or enabled via a feature flag for use by the VW standalone program.  This gives the best of both worlds.  To make it even better, I would create a second function (just for use in C libraries and programs) that reverts the emission of errors back to signals.

What does everyone think?

/cc @jmorra

**Comment by deaktator** [comment]
*BTW*, I don't think it's really important what causes the signals / exceptions.  I think the important thing is determining the type of signals from which recovery should be possible and during what stages (*initialization*, *prediction*, *updating*) recovery should be possible.

**Comment by JohnLangford** [comment]
It seems --signal_to_exception might be the right way to handle this in a
global manner.  It would be a rarely used flag which can be invoked for a
library interface.

It might also be possible to harden the file-reading code in this
particular instance, but I need time to do so---still haven't caught with
other VW stuff yet.

-John

**Comment by deaktator** [comment]
I agree.  I think this would allow both backward compatibility and the recoverability that we want.

**Comment by JohnLangford** [comment]
This pull request: https://github.com/JohnLangford/vowpal_wabbit/pull/757

is at least partially addressing the issue.

**Comment by JohnLangford** [comment]
Closing this for now as the model should be bulletproof.  I still think a signal handler is a good idea.
